tests/python/pants_test/util/test_objects.py

```diff
@@ -20,12 +23,57 @@ class AbsClass(object):
   pass
 
 
+class SomeTypedDatatype(typed_datatype('SomeTypedDatatype', {'my_val': int})):
+  pass
+
+
+class AnotherTypedDatatype(typed_datatype(str('AnotherTypedDatatype'), {
+    'a_field': str,
+    'better_field': list,
+})):
+  pass
+
+
+# TODO(cosmicexplorer): Could we have a more concise syntax for narrowing values
+# acceptable as args in a type using a predicate?
+class NonNegativeInt(typed_datatype('NonNegativeInt', {
+    'value': int,
+})):
+
+  def __new__(cls, value):
+    result = super(NonNegativeInt, cls).__new__(cls, value=value)
+
+    if result.value < 0:
+      raise TypeCheckError(
+        'NonNegativeInt', "value is negative: '{}'".format(value))
+
+    return result
+
+
+StrOrInt = SimpleTypeDecl(int).compose(SimpleTypeDecl(str))
+
+
+class YetAnotherNamedTypedDatatype(typed_datatype(
+    'YetAnotherNamedTypedDatatype', {
+      'nothing_special': str,
+      'just_another_arg': StrOrInt,
```

Comment:
You could also create an explicit union here.

Reply:
That's true, and if I wanted to use a field like `StrOrInt`, I think that would be *much* more appropriate as its own `typed_datatype` (which could then have its own set of `@rule`s converting it into some other type, if necessary.

Reply:
Using [`Exactly.from_type_or_collection()`](https://github.com/pantsbuild/pants/blob/ae70e028fd14a12ab3fbe10f9ceb0c92eea6027c/src/python/pants/util/objects.py#L193), I [converted this into `[int, str]`](https://github.com/pantsbuild/pants/blob/ae70e028fd14a12ab3fbe10f9ceb0c92eea6027c/tests/python/pants_test/util/test_objects.py#L161), which is then converted into an `Exactly` type constraint.